File: frontend/lib/api/plan.ts
// Typed REAL-mode client for platform plan/plan-feature administration
// (backend/app/api/v1/platform.py, M5). Requires PLATFORM_MANAGE on the
// backend — a non-platform-admin user calling these gets a 403 regardless of
// what the frontend shows. This client is intentionally thin: it only maps
// HTTP requests/responses to typed shapes. All validation/audit/transaction
// behavior lives in app.services.plan_service (M5) — nothing here duplicates it.

import { apiRequest } from "@/lib/apiClient";
import type { AuditEventResponse } from "@/lib/api/audit";

export interface PlanResponse {
  id: string;
  name: string;
  code: string;
  description: string | null;
  is_active: boolean;
  suite_id?: string | null;
  asset_scope?: string | null;
  included_features_count?: number;
  tenant_count?: number;
  created_at: string;
  updated_at: string;
}

export interface PlanFeatureResponse {
  id: string;
  plan_id: string;
  feature_key: string;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface PlanCreateRequest {
  name: string;
  code: string;
  description?: string | null;
  is_active?: boolean;
  suite_id?: string | null;
  asset_scope?: string | null;
}

export interface PlanUpdateRequest {
  name?: string;
  code?: string;
  description?: string | null;
  suite_id?: string | null;
  asset_scope?: string | null;
}

export interface PlanFeatureCreateRequest {
  feature_key: string;
  enabled?: boolean;
}

export interface PlanFeatureBulkItem {
  feature_key: string;
  enabled: boolean;
}

export interface PlanSubscribedTenantResponse {
  organization_id: string;
  organization_name: string;
  organization_status: string;
  subscription_id: string;
  subscription_status: string;
  starts_at: string;
  ends_at: string | null;
}

export interface PlanLimitItem {
  limit_key: string;
  limit_value: number | null;
  is_unlimited: boolean;
}

export interface PlanLimitResponse {
  id: string;
  plan_id: string;
  limit_key: string;
  limit_value: number | null;
  is_unlimited: boolean;
  created_at: string;
  updated_at: string;
}

export const planApi = {
  listPlans: (accessToken: string) => apiRequest<PlanResponse[]>("/platform/plans", { accessToken }),

  createPlan: (accessToken: string, payload: PlanCreateRequest) =>
    apiRequest<PlanResponse>("/platform/plans", { method: "POST", body: payload, accessToken }),

  getPlan: (accessToken: string, planId: string) =>
    apiRequest<PlanResponse>(`/platform/plans/${planId}`, { accessToken }),

  updatePlan: (accessToken: string, planId: string, payload: PlanUpdateRequest) =>
    apiRequest<PlanResponse>(`/platform/plans/${planId}`, { method: "PATCH", body: payload, accessToken }),

  activatePlan: (accessToken: string, planId: string) =>
    apiRequest<PlanResponse>(`/platform/plans/${planId}/activate`, { method: "POST", accessToken }),

  deactivatePlan: (accessToken: string, planId: string) =>
    apiRequest<PlanResponse>(`/platform/plans/${planId}/deactivate`, { method: "POST", accessToken }),

  listPlanFeatures: (accessToken: string, planId: string) =>
    apiRequest<PlanFeatureResponse[]>(`/platform/plans/${planId}/features`, { accessToken }),

  createPlanFeature: (accessToken: string, planId: string, payload: PlanFeatureCreateRequest) =>
    apiRequest<PlanFeatureResponse>(`/platform/plans/${planId}/features`, {
      method: "POST",
      body: payload,
      accessToken,
    }),

  bulkSetPlanFeatures: (accessToken: string, planId: string, features: PlanFeatureBulkItem[]) =>
    apiRequest<PlanFeatureResponse[]>(`/platform/plans/${planId}/features`, {
      method: "PUT",
      body: { features },
      accessToken,
    }),

  setPlanFeatureEnabled: (accessToken: string, planId: string, featureKey: string, enabled: boolean) =>
    apiRequest<PlanFeatureResponse>(`/platform/plans/${planId}/features/${encodeURIComponent(featureKey)}`, {
      method: "PATCH",
      body: { enabled },
      accessToken,
    }),

  listPlanLimits: (accessToken: string, planId: string) =>
    apiRequest<PlanLimitResponse[]>(`/platform/plans/${planId}/limits`, { accessToken }),

  bulkSetPlanLimits: (accessToken: string, planId: string, limits: PlanLimitItem[]) =>
    apiRequest<PlanLimitResponse[]>(`/platform/plans/${planId}/limits`, {
      method: "PUT",
      body: { limits },
      accessToken,
    }),

  listPlanTenants: (accessToken: string, planId: string) =>
    apiRequest<PlanSubscribedTenantResponse[]>(`/platform/plans/${planId}/tenants`, { accessToken }),

  getPlanAuditTrail: (accessToken: string, planId: string) =>
    apiRequest<AuditEventResponse[]>(`/platform/plans/${planId}/audit-trail`, { accessToken }),
};

File: backend/app/schemas/plan.py
"""M5: request/response schemas for plan and plan-feature administration.

Mirrors the style of app/schemas/platform.py and app/schemas/entitlement.py --
pure serialization/validation shapes, no business logic. ORM objects
(``Plan``, ``PlanFeature``) are never returned directly from the API layer.
"""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator


class PlanCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    code: str = Field(min_length=1, max_length=64)
    description: str | None = None
    is_active: bool = True
    suite_id: uuid.UUID | None = None
    asset_scope: str | None = Field(default=None, max_length=32)


class PlanUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    code: str | None = Field(default=None, min_length=1, max_length=64)
    description: str | None = None
    suite_id: uuid.UUID | None = None
    asset_scope: str | None = Field(default=None, max_length=32)


class PlanResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    code: str
    description: str | None
    is_active: bool
    suite_id: uuid.UUID | None = None
    asset_scope: str | None = None
    included_features_count: int = 0
    tenant_count: int = 0
    created_at: datetime
    updated_at: datetime


class PlanFeatureCreateRequest(BaseModel):
    feature_key: str = Field(min_length=1, max_length=128)
    enabled: bool = True


class PlanFeatureUpdateRequest(BaseModel):
    enabled: bool


class PlanFeatureBulkItem(BaseModel):
    feature_key: str = Field(min_length=1, max_length=128)
    enabled: bool = True


class PlanFeatureBulkUpdateRequest(BaseModel):
    features: list[PlanFeatureBulkItem]


class PlanFeatureResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plan_id: uuid.UUID
    feature_key: str
    enabled: bool
    created_at: datetime
    updated_at: datetime



class PlanSubscribedTenantResponse(BaseModel):
    organization_id: uuid.UUID
    organization_name: str
    organization_status: str
    subscription_id: uuid.UUID
    subscription_status: str
    starts_at: datetime
    ends_at: datetime | None = None


class PlanLimitItem(BaseModel):
    limit_key: str = Field(min_length=1, max_length=128)
    limit_value: int | None = Field(default=None, ge=0)
    is_unlimited: bool = False

    @model_validator(mode="after")
    def validate_limits(self) -> "PlanLimitItem":
        if self.is_unlimited:
            self.limit_value = None
        elif self.limit_value is None:
            raise ValueError("limit_value is required when is_unlimited is False")
        return self


class PlanLimitBulkUpdateRequest(BaseModel):
    limits: list[PlanLimitItem]


class PlanLimitResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plan_id: uuid.UUID
    limit_key: str
    limit_value: int | None
    is_unlimited: bool
    created_at: datetime
    updated_at: datetime


